src/compiler/scala/tools/nsc/typechecker/Implicits.scala

```diff
@@ -572,6 +572,41 @@ trait Implicits {
        }
      }
 
+    private def matchesPtInst(info: ImplicitInfo): Boolean = {
+      if (StatisticsStatics.areSomeColdStatsEnabled) statistics.incCounter(matchesPtInstCalls)
+      info.tpe match {
+        case PolyType(tparams, restpe) =>
+          val tp = ApproximateDependentMap(restpe)
+          val allUndetparams = (undetParams ++ tparams).distinct
+          try {
+            val tvars = allUndetparams map freshVar
+            val tpInstantiated = tp.instantiateTypeParams(allUndetparams, tvars)
+            if(!matchesPt(tpInstantiated, wildPt, allUndetparams)) {
+              if (StatisticsStatics.areSomeColdStatsEnabled) statistics.incCounter(matchesPtInstMismatch1)
+              false
+            } else {
+              tvars.foreach(_.constr.stopWideningIfPrecluded)
+              val targs = solvedTypes(tvars, allUndetparams, allUndetparams map varianceInType(wildPt), upper = false, lubDepth(tpInstantiated :: wildPt :: Nil))
+              val AdjustedTypeArgs(okParams, okArgs) = adjustTypeArgs(allUndetparams, tvars, targs)
+              if (!checkBounds(EmptyTree, NoPrefix, NoSymbol, okParams, okArgs, "inferred ")) {
+                if (StatisticsStatics.areSomeColdStatsEnabled) statistics.incCounter(matchesPtInstMismatch2)
+                false
+              } else {
```

Comment:
Since the third pruning seems to yield little benefit in the stats you shared, I'd consider dropping it. The whole check could then be simplified to 

```
matchesPt(tp.instantiateTypeParams(allUndetparams, tvars), wildPt, allUndetparams) && {
  tvars.foreach(_.constr.stopWideningIfPrecluded)
  val targs = solvedTypes(tvars, allUndetparams, allUndetparams map varianceInType(wildPt), upper = false, lubDepth(tp :: wildPt :: Nil))
  val AdjustedTypeArgs(okParams, okArgs) = adjustTypeArgs(allUndetparams, tvars, targs)
  checkBounds(EmptyTree, NoPrefix, NoSymbol, okParams, okArgs, "inferred ")
}
```

It's too bad we don't already have a method for this. Several are close, but not close enough :-)